src/middleware.js

```diff
@@ -15,6 +14,160 @@ import { handleCarBlock } from './handlers/car-block.js'
  * @typedef {import('@web3-storage/gateway-lib').UnixfsContext} UnixfsContext
  */
 
+/**
+ *
+ * @param {string} s
+ * @returns {import('./bindings.js').TokenMetadata}
+ */
+function deserializeTokenMetadata (s) {
+  // TODO should this be dag-json?
+  return JSON.parse(s)
+}
+
+/**
+ *
+ * @param {import('./bindings.js').TokenMetadata} m
+ * @returns string
+ */
+function serializeTokenMetadata (m) {
+  // TODO should this be dag-json?
+  return JSON.stringify(m)
+}
+
+/**
+ *
+ * @param {Environment} env
+ * @param {import('@web3-storage/gateway-lib/handlers').CID} cid
+ * @returns {Promise<import('./constants.js').RATE_LIMIT_EXCEEDED>}
+ */
+async function checkRateLimitForCID (env, cid) {
+  const rateLimiter = env.RATE_LIMITER
+  if (!rateLimiter) {
+    console.warn('no rate limiter found')
+    return RATE_LIMIT_EXCEEDED.NO
+  }
+  const rateLimitResponse = await rateLimiter.limit({ key: cid.toString() })
+  if (rateLimitResponse.success) {
+    return RATE_LIMIT_EXCEEDED.NO
+  } else {
+    console.log(`limiting CID ${cid}`)
+    return RATE_LIMIT_EXCEEDED.YES
+  }
+}
+
+/**
+ *
+ * @param {Environment} env
+ * @param {string} authToken
+ * @returns TokenMetadata
+ */
+async function getTokenMetadata (env, authToken) {
+  const cachedValue = await env.AUTH_TOKEN_METADATA.get(authToken)
+  // TODO: we should implement an SWR pattern here - record an expiry in the metadata and if the expiry has passed, re-validate the cache after
+  // returning the value
+  if (cachedValue) {
+    return deserializeTokenMetadata(cachedValue)
+  } else {
+    const accounting = Accounting.create({ serviceURL: env.ACCOUNTING_SERVICE_URL })
+    const tokenMetadata = await accounting.getTokenMetadata(authToken)
+    if (tokenMetadata) {
+      await env.AUTH_TOKEN_METADATA.put(authToken, serializeTokenMetadata(tokenMetadata))
```

Comment:
Please take a look at: https://developers.cloudflare.com/workers/runtime-apis/context/#waituntil

Do we need to wait for this as part of the request?

Reply:
Good catch. Yes, I believe we will need to use the `ctx.waitUntil` call here.